Add Eval newtype Exec

LukaJCB opened the pull request:
Follow up to #177. See also #170 

Pretty barebones so far, but seems to work fine :) 
The name is super horrible, but I couldn't find any good one and needed one to implement it so bikeshed away!

---

This PR adds the `Exec` data type which is a newtype for `Eval`.
Unlike `Eval` it is designed to handle side effects first and foremost, therefore it does not come with any instances that would violate this prinicipal. It's not capable of error handling, so it can't form `MonadError` or `Sync` and for this reason the `apply` method on its companion object won't return an `Exec[A]`, but an `Exec[Either[Throwable, A]]` instead, catching non fatal exceptions inside `Either`. If you're 100% sure your side effectful code doesn't throw, you can use `Exec.delayNoCatch`. `Exec` also doesn't support the `Later` constructor of `Eval`, as it would memoize the value and can cause loss of RT.

When done, this should replace the `EitherT[Eval, Throwable, A]` instance for `Sync` we currently have, as that one can lead to subtle bugs with unexpected side effects.

alexandru commented on core/shared/src/main/scala/cats/effect/EvalEff.scala:
```diff
@@ -0,0 +1,65 @@
+  def unsafeToEitherT: EitherT[Eval, Throwable, A] =
+    EitherT(unsafeToEval)
+
+  def unsafeToEval: Eval[Either[Throwable, A]] =
```
An operation that at runtime is literally a no-op cannot be “unsafe”.

alexandru commented on core/shared/src/main/scala/cats/effect/internals/EitherEvalNewtype.scala:
```diff
@@ -0,0 +1,31 @@
+  trait Tag extends Any
+  type Type[+A] <: Base with Tag
+
+  def apply[A](fa: Eval[Either[Throwable, A]]): Type[A] =
```
We’re defining a newtype for EitherT[Eval], but this raises the question, why not an alias for Eval, with EitherT out of the picture?

After all, as I’ve been saying all along, Eval is equipped to handle Sync already.

LukaJCB replied:
You're right, I like that idea more!

LukaJCB commented:
Exec is great! :+1:

johnynek commented on core/shared/src/main/scala/cats/effect/Exec.scala:
```diff
@@ -0,0 +1,108 @@
+    * instances will be converted into thunk-less `Exec` (i.e. eager
+    * `Exec`), while lazy eval and memoized will be executed as such.
+    */
+  def eval[A](fa: Eval[A]): IO[A] = fa match {
```
This should return Exec, no?

alexandru commented on core/shared/src/main/scala/cats/effect/package.scala:
```diff
@@ -0,0 +1,29 @@
+package object effect {
+
+  /**
+    * Exec is a data type that allows for suspending synchronous computations.
```
Minor style nitpick, but in `cats-effect`, like in Cats, we've been using the old JavaDoc-style comments. You can set project specific IntelliJ IDEA settings btw.

LukaJCB replied:
Ah, I missed that, will fix!

LukaJCB commented:
@johnynek, I agree we definitely need a newtype solution for `EitherT`, but I'd rather keep the two issues separate for now.

alexandru commented on laws/shared/src/main/scala/cats/effect/laws/util/TestInstances.scala:
```diff
@@ -88,6 +88,9 @@ trait TestInstances {
     }
+
+  implicit def eqExec[A: Eq]: Eq[Exec[A]] =
+    Eq.by(_.unsafeRun)
```
`unsafeRun` can throw. And the error should be caught in a `Try` or `Either` or something.

johnynek replied:
how can it throw? I thought the whole point of `Exec` is that it throws only fatal errors.

alexandru replied:
Ignore my comment, I assumed that we were implementing Sync[Exec], but apparently we can’t do that.

alexandru commented on laws/shared/src/main/scala/cats/effect/laws/discipline/Arbitrary.scala:
```diff
@@ -24,11 +24,15 @@ import cats.effect.internals.IORunLoop
     Arbitrary(Gen.delay(genIO[A]))
 
+  implicit def catsEffectLawsArbitraryForExec[A: Arbitrary]: Arbitrary[Exec[A]] =
+    Arbitrary(catsLawsArbitraryForEval[A].arbitrary.map(e => Exec.create(e)))
```
This is insufficient as this `Arbitrary[Exec]` instance also needs to yield `Exec` values that terminate in error.

alexandru replied:
Ignore the comment above.

alexandru commented on core/shared/src/main/scala/cats/effect/Sync.scala:
```diff
@@ -87,6 +87,34 @@ object Sync {
       }
     }
 
+  implicit val catsEitherTEvalEffSync: Sync[EitherT[Exec, Throwable, ?]] =
```
Ah, OK, this is something I missed.

Question: why not add a `Sync[Exec]`?

After all, it's possible:
https://github.com/typelevel/cats/pull/1558

alexandru replied:
@johnynek I'm interested in your opinion as you were also involved in that PR I linked above.

Anything wrong with implementing `MonadError` for `Exec`, like Erik did in that PR?

LukaJCB replied:
Related https://github.com/typelevel/cats/issues/1639

LukaJCB replied:
Also https://github.com/typelevel/cats/issues/1661

edmundnoble commented on core/shared/src/main/scala/cats/effect/package.scala:
```diff
@@ -0,0 +1,29 @@
+    * and in order to form a valid `Sync` needs to be combined with `EitherT`.
+    */
+  type Exec[A] = ExecImpl.Type[A]
+  val Exec = ExecImpl
```
This code is a mix of multiple newtype encodings at once. If you aren't going to use the abstract type trick to avoid writing casts in `ExecNewtype` - and I would recommend that if you don't, you still at least provide a `def subst[F[_[_]]](fs: F[Eval]): F[Exec]` to give you zero-cost conversions) - I would say just use `object Exec`.

LukaJCB commented:
I split the prior implementation into an unexceptional and exceptional one. Seems like `attempt` isn't stack safe yet, will need some extra work.

johnynek commented on core/shared/src/main/scala/cats/effect/internals/ExecNewtype.scala:
```diff
@@ -0,0 +1,31 @@
+    fa.asInstanceOf[Type[A]]
+
+  private[effect] def unwrap[A](fa: Type[A]): Eval[A] =
+    fa.asInstanceOf[Eval[A]]
```
if we had `subst[F[_[_]]](F[Exec]): F[Eval]` we could use it to cast `A => Exec[B]` into  `A => Eval[B]` with zero cost. This would be nice to have below where you are using `unwrap` inside andThen which is just wasting a call to a cast. Having subst in both directions would allow us to only ever cast and never using create or effect on higher kinded things.

LukaJCB replied:
Good call, will do :)

alexandru commented on core/shared/src/main/scala/cats/effect/Exec.scala:
```diff
@@ -0,0 +1,151 @@
+import cats.effect.internals.{ExecNewtype, NonFatal}
+import cats.kernel.{Monoid, Semigroup}
+
+object ExecImpl extends ExecInstances with ExecNewtype {
```
I don't like the idea of exposing an `ExecImpl` in the `cats.effect` namespace.

Can't we work directly with `object Exec`, instead of `val Exec = ExecImpl`?
I think we can.

alexandru commented on core/shared/src/main/scala/cats/effect/Exec.scala:
```diff
@@ -0,0 +1,151 @@
+  /**
+    * Suspends a synchronous side effect in `Exec` and catches non-fatal exceptions.
+    */
+  def apply[A](thunk: => A): Exec[A] =
```
Create a shortcut for `suspend` as well.

alexandru commented on core/shared/src/main/scala/cats/effect/Exec.scala:
```diff
@@ -0,0 +1,151 @@
+  }
+
+  implicit def catsEvalEffOps[A](value: Exec[A]): ExecOps[A] =
+    new ExecOps(value)
```
This is inefficient. We can work with implicit classes, which would work if this where an `object Exec`.

alexandru commented on core/shared/src/main/scala/cats/effect/Exec.scala:
```diff
@@ -0,0 +1,151 @@
+    * Convert this `Exec` to another effect type that implements the [[Sync]] type class.
+    */
+  def to[F[_]: Sync]: F[A] =
+    Sync[F].delay(unsafeRun)
```
We should treat `Now` with a special case imo, initializing it with `pure`.

alexandru commented on core/shared/src/main/scala/cats/effect/Exec.scala:
```diff
@@ -0,0 +1,151 @@
+  /**
+    * Convert this `Exec` to another effect type that implements the [[Sync]] type class.
+    */
+  def to[F[_]: Sync]: F[A] =
```
I don't like the `F[_]: Sync` notation for one reason: it produces garbage in ScalaDoc.

The definition for `IO` uses the `implicit F` convention and I like that better, plus it would be great if we maintained consistency:

```scala
  final def to[F[_]](implicit F: LiftIO[F]): F[A @uncheckedVariance] =
    F.liftIO(this)
```

alexandru commented on core/shared/src/main/scala/cats/effect/Exec.scala:
```diff
@@ -0,0 +1,151 @@
+    new ExecOps(value)
+}
+
+sealed class ExecOps[A](val value: Exec[A]) {
```
`Eval`'s type parameter is covariant, `IO`'s type parameter is covariant as well.

`Exec` should have a covariant type parameter too.

alexandru commented on core/shared/src/main/scala/cats/effect/UExec.scala:
```diff
@@ -0,0 +1,129 @@
+import cats.effect.internals.{ExecNewtype, NonFatal}
+import cats.kernel.{Monoid, Semigroup}
+
+object UExecImpl extends UExecInstances with ExecNewtype {
```
I am not sold on the separation between `Exec` and `UExec`.

I know we're talking of #176, however that's a lot of effort, we haven't decided what to do and I'm wary of polluting the namespace.

Get rid of `UExec` and `Exec` should not implement `MonadError` directly because it's apparently problematic — if we can't make `handleErrorWith` stack safe, then we can't do it.

(unless you found a way to describe a stack safe `handleErrorWith`)

LukaJCB replied:
I'm inclined to agree, I haven't found a way to make it stack safe, though I wonder if we could make `attempt` stack safe instead. I need to think about this more when there's a quiet moment to concentrate. 
If anyone has any pointers, I'd be glad to hear them :smile:

alexandru commented on core/shared/src/main/scala/cats/effect/package.scala:
```diff
@@ -0,0 +1,32 @@
+  val UExec = UExecImpl
+
+  type Exec[A] = ExecImpl.Type[A]
+  val Exec = ExecImpl
```
Reiterating what I mentioned above, I think we should not work with `val Exec`.

alexandru commented on site/src/main/tut/datatypes/io.md:
```diff
@@ -61,7 +61,7 @@ it is sequenced in the monadic chain.
 |:------------------:|:-------------------:|:--------------------------:|
 | **Synchronous**    |          A          |           () => A          |
-|                    |                     | [Eval[A]](https://typelevel.org/cats/datatypes/eval.html) |
+|                    |                     | [UExec[A]](UExec.html)       |
```
In terms of the evaluation model, `Eval` is still the equivalent of a thunk.

LukaJCB commented:
I've removed the `Sync[Exec]` instance since it wasn't stack safe and went back to `EitherT[Exec, Throwable, A]` and addressed most of the feedback :)

LukaJCB commented:
Unfortunately just using one `object Exec` does not work on 2.11, so I had to go back to `object ExecImpl` instead. :/

I did make it package private though to avoid polluting the name space.

johnynek commented on core/shared/src/main/scala/cats/effect/Exec.scala:
```diff
@@ -0,0 +1,130 @@
+    * `Exec`), while lazy eval and memoized will be executed as such.
+    */
+  def eval[A](fa: Eval[A]): Exec[A] = fa match {
+    case Now(a) => pure(a)
```
why is this not `create(fa)`?

LukaJCB replied:
Oversight, good point, thanks! :)

alexandru commented on core/shared/src/main/scala/cats/effect/Exec.scala:
```diff
@@ -0,0 +1,130 @@
+import cats.effect.internals.{ExecNewtype, NonFatal}
+import cats.kernel.{Monoid, Semigroup}
+
+private[effect] object ExecImpl extends ExecInstances with ExecNewtype {
```
It is weird that this can be private, as you can’t expose package private types in public APIs. Doesn’t the compiler throw an error?

You should double check what actually gets exported via `val Exec = ExecImpl`.

LukaJCB replied:
It is weird indeed, but it seems to work fine in both the REPL and the tests and I'm not sure how to further Double Check 🤔

oleg-py commented:
@LukaJCB maybe it's better then to make Exec a newtype to `EitherT[Eval, Throwable, ?]`? It feels odd that we have a type intended for capturing effects but has no instance of cats-effect typeclass on its own.

oleg-py commented:
I just had a dank idea: make `Exec` a newtype of `IO` instead. Re-implement bracket without `onCancelRaiseError` and provide only `.unsafeRunSync()`.

alexandru commented:
I think we should defer this one post 1.0 at least.

For one, I don't like newtypes due to not having a standard encoding and due to general awkwardness, that might get fixed with opaque types, although a newtype for `IO` might be a good idea.

Also, if we end up with a `SyncIO`, then it will make sense to replace our usage of `IO` in the type classes (e.g. `Concurrent`, `Effect` and `ConcurrentEffect`, so whenever we used `IO[Unit]` instead of `() => Unit`).

mpilquist commented:
I'm interested in an `Exec` (or `SyncIO`) wrapper around `IO` that only exposes sync constructors. Such a type could then at least replace `IO` in the hierarchy. I don't mind getting a PR together if there are no objections, though my goal would be to have this in before 1.0.

LukaJCB commented:
Closing in favor of #297 :)
